File: yandex_training/1_7/D.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

int main() {
  int n;
  std::cin >> n;

  std::vector<std::tuple<int, int, int>> events;

  for (int i = 0; i < n; ++i) {
    int nowin, nowout;
    std::cin >> nowin >> nowout;

    if (nowout - nowin < 5) {
      continue;
    }

    events.emplace_back(nowin, -1, i);
    events.emplace_back(nowout - 5, 1, i);
  }

  sort(events.begin(), events.end());

  if (events.empty()) {
    std::cout << "0 10 20" << std::endl;
  } else if (events.size() == 2) {
    std::cout << 1 << " " << std::get<0>(events[0]) << " "
              << std::get<0>(events[0]) + 10 << std::endl;
  } else {
    int best = 0;
    int first_best = 0, second_best = 0;
    std::set<int> first_ad;

    for (size_t i = 0; i < events.size(); ++i) {
      auto event1 = events[i];
      if (std::get<1>(event1) == -1) {
        first_ad.insert(std::get<2>(event1));
        if (first_ad.size() > best) {
          best = first_ad.size();
          first_best = std::get<0>(event1);
          second_best = std::get<0>(event1) + 5;
        }
      }

      int second_ad_cnt = 0;

      for (size_t j = i + 1; j < events.size(); ++j) {
        auto event2 = events[j];
        if (std::get<1>(event2) == -1 &&
            first_ad.find(std::get<2>(event2)) == first_ad.end()) {
          second_ad_cnt++;
        }
        if (std::get<0>(event2) - 5 >= std::get<0>(event1) &&
            first_ad.size() + second_ad_cnt > best) {
          best = first_ad.size() + second_ad_cnt;
          first_best = std::get<0>(event1);
          second_best = std::get<0>(event2);
        }
        if (std::get<1>(event2) == 1 &&
            first_ad.find(std::get<2>(event2)) == first_ad.end()) {
          second_ad_cnt--;
        }
      }

      if (std::get<1>(event1) == 1) {
        first_ad.erase(std::get<2>(event1));
      }
    }

    std::cout << best << " " << first_best << " " << second_best << std::endl;
  }

  return 0;
}

File: yandex_training/1_7/D.py
n = int(input())
events: list[tuple[int, int, int]] = []

for i in range(n):
    nowin, nowout = map(int, input().split())
    if nowout - nowin < 5:
        continue

    events.append((nowin, -1, i))
    events.append((nowout - 5, 1, i))

events.sort()

if len(events) == 0:
    print("0 10 20")
elif len(events) == 2:
    print(1, events[0][0], events[0][0] + 10)
else:
    best = 0
    first_best, second_best = 0, 0
    first_ad: set[int] = set()

    for i in range(len(events)):
        event1 = events[i]
        if event1[1] == -1:
            first_ad.add(event1[2])
            if len(first_ad) > best:
                best = len(first_ad)
                first_best = event1[0]
                second_best = event1[0] + 5

        second_ad_cnt = 0

        for j in range(i + 1, len(events)):
            event2 = events[j]
            if event2[1] == -1 and event2[2] not in first_ad:
                second_ad_cnt += 1
            if event2[0] - 5 >= event1[0] and len(first_ad) + second_ad_cnt > best:
                best = len(first_ad) + second_ad_cnt
                first_best = event1[0]
                second_best = event2[0]
            if event2[1] == 1 and event2[2] not in first_ad:
                second_ad_cnt -= 1

        if event1[1] == 1:
            first_ad.remove(event1[2])

    print(best, first_best, second_best)
